Messages aren't easy to understand

Hello,

I think messages returned for failed cases aren't very easy.

Here is an example:

```java
  @Property
  public void firstTest(final List<Integer>values) {
    final ExerciseOne instance = new ExerciseOne();
    Assert.assertTrue(instance.absoluteAverage(values) >= 0);
  }
```

Here is the message I get:

```java
java.lang.AssertionError: Property firstTest falsified via shrinking: null
Shrunken args: [[-1209279219, 938232675]]
Original failure message: [null]
Original args: [[-1209279219, -1986811869]]
Seeds: [7190051195178127153]
```

Here, I thought that the failing case is the input `null`, but it's not the case, as this check is manually checked in another case.

Instead, I feel like it's "[-1209279219, 938232675]", even though I thought the lib tried to provide a list of list of integers at first (it's an overflow issue here).

Could the failing case be clearer from the log messages?

For example, what about:

``` java
java.lang.AssertionError: Property named 'firstTest' failed with the following shrunk arguments: [-1209279219, 938232675]
First value found to also provoke a failure: [-1209279219, -1986811869]
Seeds: [7190051195178127153]
```

Version tested: 0.7 & 0.8-beta-2.

--- a/core/src/main/java/com/pholser/junit/quickcheck/runner/PropertyFalsified.java
+++ b/core/src/main/java/com/pholser/junit/quickcheck/runner/PropertyFalsified.java
@@ -37,10 +37,10 @@ final class PropertyFalsified {
         Object[] args,
         long[] seeds,
         AssertionError e) {
-        final String message = "Property named '%s' failed%s%n"
+
+        String message = "Property named '%s' failed%s%n"
             + "With arguments: %s%n"
             + "Seeds for reproduction: %s";
-
 
         return new AssertionError(
             String.format(
@@ -59,7 +59,8 @@ final class PropertyFalsified {
         long[] seeds,
         AssertionError smallerFailure,
         AssertionError originalFailure) {
-        final String message = "Property named '%s' failed%s%n"
+
+        String message = "Property named '%s' failed%s%n"
             + "With arguments: %s%n"
             + (originalFailure.getMessage() == null ? ""
                 : "Original failure message: " + originalFailure.getMessage() + "%n")

--- a/core/src/main/java/com/pholser/junit/quickcheck/runner/PropertyFalsified.java
+++ b/core/src/main/java/com/pholser/junit/quickcheck/runner/PropertyFalsified.java
@@ -37,14 +37,16 @@ final class PropertyFalsified {
         Object[] args,
         long[] seeds,
         AssertionError e) {
+        final String message = "Property named '%s' failed%s%n"
+            + "With arguments: %s%n"
+            + "Seeds for reproduction: %s";
+
 
         return new AssertionError(
             String.format(
-                "Property %s falsified.%s%n"
-                    + "Args: %s%n"
-                    + "Seeds: %s%n",
+                message,
                 propertyName,
-                e.getMessage(),
+                e.getMessage() == null ? ":" : " (" + e.getMessage() + "):",
                 Arrays.deepToString(args),
                 Arrays.toString(seeds)),
             e);
@@ -57,18 +59,20 @@ final class PropertyFalsified {
         long[] seeds,
         AssertionError smallerFailure,
         AssertionError originalFailure) {
+        final String message = "Property named '%s' failed%s%n"
+            + "With arguments: %s%n"
+            + (originalFailure.getMessage() == null ? ""
+                : "Original failure message: " + originalFailure.getMessage() + "%n")
+            + "First arguments found to also provoke a failure: %s%n"
+            + "Seeds for reproduction: %s";
 
         AssertionError e = new AssertionError(
             String.format(
-                "Property %s falsified via shrinking: %s%n"
-                    + "Shrunken args: %s%n"
-                    + "Original failure message: [%s]%n"
-                    + "Original args: %s%n"
-                    + "Seeds: %s%n",
+                message,
                 propertyName,
-                smallerFailure.getMessage(),
+                smallerFailure.getMessage() == null ? ":"
+                    : " (" + smallerFailure.getMessage() + "):",
                 Arrays.deepToString(args),
-                originalFailure.getMessage(),
                 Arrays.deepToString(originalArgs),
                 Arrays.toString(seeds)),
             originalFailure);

--- a/core/src/test/java/com/pholser/junit/quickcheck/ShrinkingTest.java
+++ b/core/src/test/java/com/pholser/junit/quickcheck/ShrinkingTest.java
@@ -43,10 +43,10 @@ import static org.junit.experimental.results.PrintableResult.*;
 import static org.junit.experimental.results.ResultMatchers.*;
 
 public class ShrinkingTest {
-    @Test public void complete() throws Exception {
+    @Test public void complete() {
         assertThat(
             testResult(ShrinkingCompletely.class),
-            hasSingleFailureContaining(String.format("Shrunken args: [%s]", new Foo(1))));
+            hasSingleFailureContaining(String.format("With arguments: [%s]", new Foo(1))));
     }
 
     @RunWith(JUnitQuickcheck.class)
@@ -59,10 +59,10 @@ public class ShrinkingTest {
         }
     }
 
-    @Test public void shrinkingDoesNotShrink() throws Exception {
+    @Test public void shrinkingDoesNotShrink() {
         assertThat(
             testResult(ShrinkingNotReally.class),
-            hasSingleFailureContaining("Args: ["));
+            hasSingleFailureContaining("With arguments: ["));
     }
 
     @RunWith(JUnitQuickcheck.class)
@@ -74,10 +74,10 @@ public class ShrinkingTest {
         }
     }
 
-    @Test public void shrinkingDoesNotShrinkWhenLargerEqualsSmaller() throws Exception {
+    @Test public void shrinkingDoesNotShrinkWhenLargerEqualsSmaller() {
         assertThat(
             testResult(ShrinksAreIdentity.class),
-            hasSingleFailureContaining("Args: ["));
+            hasSingleFailureContaining("With arguments: ["));
     }
 
     @RunWith(JUnitQuickcheck.class)
@@ -90,7 +90,7 @@ public class ShrinkingTest {
     @Test public void assumptionFailureWhileShrinking() {
         assertThat(
             testResult(FailedAssumptionDuringShrinking.class),
-            hasSingleFailureContaining("Args: ["));
+            hasSingleFailureContaining("With arguments: ["));
     }
 
     @RunWith(JUnitQuickcheck.class)
@@ -207,7 +207,7 @@ public class ShrinkingTest {
         assertThat(
             testResult(ShrinkingMoreThanOnePropertyParameter.class),
             hasSingleFailureContaining(
-                String.format("Shrunken args: [%s, %s]", new Foo(1), new Foo(1))));
+                String.format("With arguments: [%s, %s]", new Foo(1), new Foo(1))));
     }
 
     @RunWith(JUnitQuickcheck.class)
@@ -221,10 +221,10 @@ public class ShrinkingTest {
         }
     }
 
-    @Test public void timeout() throws Exception {
+    @Test public void timeout() {
         assertThat(
             testResult(ShrinkingTimeout.class),
-            hasSingleFailureContaining("Args: ["));
+            hasSingleFailureContaining("With arguments: ["));
     }
 
     @RunWith(JUnitQuickcheck.class)

--- a/core/src/test/java/com/pholser/junit/quickcheck/runner/PropertyFalsifiedUtilityClassTest.java
+++ b/core/src/test/java/com/pholser/junit/quickcheck/runner/PropertyFalsifiedUtilityClassTest.java
@@ -26,11 +26,119 @@
 package com.pholser.junit.quickcheck.runner;
 
 import com.pholser.junit.quickcheck.UtilityClassesUninstantiabilityHarness;
+import org.junit.*;
+
 
 public class PropertyFalsifiedUtilityClassTest
     extends UtilityClassesUninstantiabilityHarness {
 
     public PropertyFalsifiedUtilityClassTest() {
         super(PropertyFalsified.class);
+    }
+
+    @Test
+    public void counterexampleFoundWithAllParametersIsCorrect() {
+        final String propertyName = "mySuperProperty";
+        final String[] arguments = {"first", "second", "third"};
+        final long[] seeds = {12345, 8842};
+        final String assertionName = "assertion name";
+        final AssertionError error = new AssertionError(assertionName);
+        final AssertionError actual = PropertyFalsified.counterexampleFound(
+            propertyName,
+            arguments,
+            seeds,
+            error);
+        final String expected = "Property named 'mySuperProperty' failed (assertion name):\n"
+            + "With arguments: [first, second, third]\n"
+            + "Seeds for reproduction: [12345, 8842]";
+        Assert.assertEquals(expected, actual.getMessage());
+    }
+
+    @Test
+    public void counterexampleFoundWhenAssertionErrorPassedHasNoMessageIsCorrect() {
+        final String propertyName = "mySuperProperty";
+        final String[] arguments = {"first", "second", "third"};
+        final long[] seeds = {12345, 8842};
+        final AssertionError error = new AssertionError();
+        final AssertionError actual = PropertyFalsified.counterexampleFound(
+            propertyName,
+            arguments,
+            seeds,
+            error);
+        final String expected = "Property named 'mySuperProperty' failed:\n"
+            + "With arguments: [first, second, third]\n"
+            + "Seeds for reproduction: [12345, 8842]";
+        Assert.assertEquals(expected, actual.getMessage());
+    }
+
+    @Test
+    public void smallerCounterexampleFoundWithAllParametersIsCorrect() {
+        final String propertyName = "mySuperProperty";
+        final String[] originalArguments = {"first", "second", "third"};
+        final String[] arguments = {"first"};
+        final long[] seeds = {12345, 8842};
+        final String smallerFailureName = "smaller name";
+        final AssertionError smallerFailure = new AssertionError(smallerFailureName);
+        final String assertionName = "assertion name";
+        final AssertionError originalFailure = new AssertionError(assertionName);
+        final AssertionError actual = PropertyFalsified.smallerCounterexampleFound(
+            propertyName,
+            originalArguments,
+            arguments,
+            seeds,
+            smallerFailure,
+            originalFailure);
+        final String expected = "Property named 'mySuperProperty' failed (smaller name):\n"
+            + "With arguments: [first]\n"
+            + "Original failure message: assertion name\n"
+            + "First arguments found to also provoke a failure: [first, second, third]\n"
+            + "Seeds for reproduction: [12345, 8842]";
+        Assert.assertEquals(expected, actual.getMessage());
+    }
+
+    @Test
+    public void smallerCounterexampleFoundIsCorrectEvenIfSmallerFailureIsNotNamed() {
+        final String propertyName = "mySuperProperty";
+        final String[] originalArguments = {"first", "second", "third"};
+        final String[] arguments = {"first"};
+        final long[] seeds = {12345, 8842};
+        final AssertionError smallerFailure = new AssertionError();
+        final String assertionName = "assertion name";
+        final AssertionError originalFailure = new AssertionError(assertionName);
+        final AssertionError actual = PropertyFalsified.smallerCounterexampleFound(
+            propertyName,
+            originalArguments,
+            arguments,
+            seeds,
+            smallerFailure,
+            originalFailure);
+        final String expected = "Property named 'mySuperProperty' failed:\n"
+            + "With arguments: [first]\n"
+            + "Original failure message: assertion name\n"
+            + "First arguments found to also provoke a failure: [first, second, third]\n"
+            + "Seeds for reproduction: [12345, 8842]";
+        Assert.assertEquals(expected, actual.getMessage());
+    }
+
+    @Test
+    public void smallerCounterexampleFoundIsCorrectEvenIfOriginalFailureIsNotNamed() {
+        final String propertyName = "mySuperProperty";
+        final String[] originalArguments = {"first", "second", "third"};
+        final String[] arguments = {"first"};
+        final long[] seeds = {12345, 8842};
+        final AssertionError smallerFailure = new AssertionError();
+        final AssertionError originalFailure = new AssertionError();
+        final AssertionError actual = PropertyFalsified.smallerCounterexampleFound(
+            propertyName,
+            originalArguments,
+            arguments,
+            seeds,
+            smallerFailure,
+            originalFailure);
+        final String expected = "Property named 'mySuperProperty' failed:\n"
+            + "With arguments: [first]\n"
+            + "First arguments found to also provoke a failure: [first, second, third]\n"
+            + "Seeds for reproduction: [12345, 8842]";
+        Assert.assertEquals(expected, actual.getMessage());
     }
 }

--- a/core/src/test/java/com/pholser/junit/quickcheck/runner/PropertyFalsifiedUtilityClassTest.java
+++ b/core/src/test/java/com/pholser/junit/quickcheck/runner/PropertyFalsifiedUtilityClassTest.java
@@ -26,7 +26,9 @@
 package com.pholser.junit.quickcheck.runner;
 
 import com.pholser.junit.quickcheck.UtilityClassesUninstantiabilityHarness;
-import org.junit.*;
+
+import org.junit.Test;
+import static org.junit.Assert.assertEquals;
 
 
 public class PropertyFalsifiedUtilityClassTest
@@ -38,107 +40,117 @@ public class PropertyFalsifiedUtilityClassTest
 
     @Test
     public void counterexampleFoundWithAllParametersIsCorrect() {
-        final String propertyName = "mySuperProperty";
-        final String[] arguments = {"first", "second", "third"};
-        final long[] seeds = {12345, 8842};
-        final String assertionName = "assertion name";
-        final AssertionError error = new AssertionError(assertionName);
-        final AssertionError actual = PropertyFalsified.counterexampleFound(
+        String propertyName = "mySuperProperty";
+        String[] arguments = {"first", "second", "third"};
+        long[] seeds = {12345, 8842};
+        String assertionName = "assertion name";
+        AssertionError error = new AssertionError(assertionName);
+
+        AssertionError actual = PropertyFalsified.counterexampleFound(
             propertyName,
             arguments,
             seeds,
             error);
-        final String expected = "Property named 'mySuperProperty' failed (assertion name):\n"
+
+        String expected = "Property named 'mySuperProperty' failed (assertion name):\n"
             + "With arguments: [first, second, third]\n"
             + "Seeds for reproduction: [12345, 8842]";
-        Assert.assertEquals(expected, actual.getMessage());
+        assertEquals(expected, actual.getMessage());
     }
 
     @Test
     public void counterexampleFoundWhenAssertionErrorPassedHasNoMessageIsCorrect() {
-        final String propertyName = "mySuperProperty";
-        final String[] arguments = {"first", "second", "third"};
-        final long[] seeds = {12345, 8842};
-        final AssertionError error = new AssertionError();
-        final AssertionError actual = PropertyFalsified.counterexampleFound(
+        String propertyName = "mySuperProperty";
+        String[] arguments = {"first", "second", "third"};
+        long[] seeds = {12345, 8842};
+        AssertionError error = new AssertionError();
+
+        AssertionError actual = PropertyFalsified.counterexampleFound(
             propertyName,
             arguments,
             seeds,
             error);
-        final String expected = "Property named 'mySuperProperty' failed:\n"
+
+        String expected = "Property named 'mySuperProperty' failed:\n"
             + "With arguments: [first, second, third]\n"
             + "Seeds for reproduction: [12345, 8842]";
-        Assert.assertEquals(expected, actual.getMessage());
+        assertEquals(expected, actual.getMessage());
     }
 
     @Test
     public void smallerCounterexampleFoundWithAllParametersIsCorrect() {
-        final String propertyName = "mySuperProperty";
-        final String[] originalArguments = {"first", "second", "third"};
-        final String[] arguments = {"first"};
-        final long[] seeds = {12345, 8842};
-        final String smallerFailureName = "smaller name";
-        final AssertionError smallerFailure = new AssertionError(smallerFailureName);
-        final String assertionName = "assertion name";
-        final AssertionError originalFailure = new AssertionError(assertionName);
-        final AssertionError actual = PropertyFalsified.smallerCounterexampleFound(
+        String propertyName = "mySuperProperty";
+        String[] originalArguments = {"first", "second", "third"};
+        String[] arguments = {"first"};
+        long[] seeds = {12345, 8842};
+        String smallerFailureName = "smaller name";
+        AssertionError smallerFailure = new AssertionError(smallerFailureName);
+        String assertionName = "assertion name";
+        AssertionError originalFailure = new AssertionError(assertionName);
+
+        AssertionError actual = PropertyFalsified.smallerCounterexampleFound(
             propertyName,
             originalArguments,
             arguments,
             seeds,
             smallerFailure,
             originalFailure);
-        final String expected = "Property named 'mySuperProperty' failed (smaller name):\n"
+
+        String expected = "Property named 'mySuperProperty' failed (smaller name):\n"
             + "With arguments: [first]\n"
             + "Original failure message: assertion name\n"
             + "First arguments found to also provoke a failure: [first, second, third]\n"
             + "Seeds for reproduction: [12345, 8842]";
-        Assert.assertEquals(expected, actual.getMessage());
+        assertEquals(expected, actual.getMessage());
     }
 
     @Test
     public void smallerCounterexampleFoundIsCorrectEvenIfSmallerFailureIsNotNamed() {
-        final String propertyName = "mySuperProperty";
-        final String[] originalArguments = {"first", "second", "third"};
-        final String[] arguments = {"first"};
-        final long[] seeds = {12345, 8842};
-        final AssertionError smallerFailure = new AssertionError();
-        final String assertionName = "assertion name";
-        final AssertionError originalFailure = new AssertionError(assertionName);
-        final AssertionError actual = PropertyFalsified.smallerCounterexampleFound(
+        String propertyName = "mySuperProperty";
+        String[] originalArguments = {"first", "second", "third"};
+        String[] arguments = {"first"};
+        long[] seeds = {12345, 8842};
+        AssertionError smallerFailure = new AssertionError();
+        String assertionName = "assertion name";
+        AssertionError originalFailure = new AssertionError(assertionName);
+
+        AssertionError actual = PropertyFalsified.smallerCounterexampleFound(
             propertyName,
             originalArguments,
             arguments,
             seeds,
             smallerFailure,
             originalFailure);
-        final String expected = "Property named 'mySuperProperty' failed:\n"
+
+        String expected = "Property named 'mySuperProperty' failed:\n"
             + "With arguments: [first]\n"
             + "Original failure message: assertion name\n"
             + "First arguments found to also provoke a failure: [first, second, third]\n"
             + "Seeds for reproduction: [12345, 8842]";
-        Assert.assertEquals(expected, actual.getMessage());
+        assertEquals(expected, actual.getMessage());
     }
 
     @Test
     public void smallerCounterexampleFoundIsCorrectEvenIfOriginalFailureIsNotNamed() {
-        final String propertyName = "mySuperProperty";
-        final String[] originalArguments = {"first", "second", "third"};
-        final String[] arguments = {"first"};
-        final long[] seeds = {12345, 8842};
-        final AssertionError smallerFailure = new AssertionError();
-        final AssertionError originalFailure = new AssertionError();
-        final AssertionError actual = PropertyFalsified.smallerCounterexampleFound(
+        String propertyName = "mySuperProperty";
+        String[] originalArguments = {"first", "second", "third"};
+        String[] arguments = {"first"};
+        long[] seeds = {12345, 8842};
+        AssertionError smallerFailure = new AssertionError();
+        AssertionError originalFailure = new AssertionError();
+
+        AssertionError actual = PropertyFalsified.smallerCounterexampleFound(
             propertyName,
             originalArguments,
             arguments,
             seeds,
             smallerFailure,
             originalFailure);
-        final String expected = "Property named 'mySuperProperty' failed:\n"
+
+        String expected = "Property named 'mySuperProperty' failed:\n"
             + "With arguments: [first]\n"
             + "First arguments found to also provoke a failure: [first, second, third]\n"
             + "Seeds for reproduction: [12345, 8842]";
-        Assert.assertEquals(expected, actual.getMessage());
+        assertEquals(expected, actual.getMessage());
     }
 }

--- a/src/site/markdown/usage/seed.md
+++ b/src/site/markdown/usage/seed.md
@@ -21,11 +21,11 @@ the failure.
 junit-quickcheck reports the seed used for a given property parameter in the
 message of the `AssertionError` raised when a property fails.
 
-    java.lang.AssertionError: Property myProperty falsified.
-    Original failure message: [
+    java.lang.AssertionError: Property named 'myProperty' failed:
+    With arguments: [10]
+    Original failure message:
     Expected: a value less than <1>
-         but: <753701363> was greater than <1>]
-    Original args: [753701363]
-    Args shrunken to: [1]
-    Seeds: [-6700838609453830748]
+         but: <10> was greater than <1>
+    First arguments found to also provoke a failure: [753701363]
+    Seeds for reproduction: [-6700838609453830748]
     ...

